File: src/1-500/140/WordBreakII.py
from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        def dfs(s, res, wordDict, memo):
            if s in memo:
                return memo[s]
            if not s:
                return [""]
            res = []
            for word in wordDict:
                if s[:len(word)] != word: continue # 如果字符串开头的单词找不到，就跳过了。
                for r in dfs(s[len(word):], res, wordDict, memo): # 如果找到了，那么剩下的字符串递归再求解。注意数组如何切分，如果没找到，不动res
                    res.append(word + ("" if not r else " " + r)) # 处理空字符串
            memo[s] = res
            return res
        res = []
        memo = {}
        return dfs(s, res, wordDict, memo)



solution = Solution()
#print(solution.wordBreak("aab", ["aa"]))
print(solution.wordBreak("pineapplepenapple", ["apple", "pen", "applepen", "pine", "pineapple"]))

File: src/1-500/140/WordBreakII.java
class Solution {
    //bottom up recursion with memorization
    public List<String> wordBreak(String s, List<String> wordDict) {
        return dfs(s, wordDict, new HashMap<String, List<String>>());
    }

    private List<String> dfs(String s, List<String> wordDict, Map<String, List<String>> memo) {
        if (memo.containsKey(s)) {
            return memo.get(s);
        }
        List<String> list = new ArrayList<>();
        if (s.length() == 0) {
            list.add("");
            return list;
        }
        for (String str : wordDict) {
            if (s.startsWith(str)) {
                List<String> subList = dfs(s.substring(str.length()), wordDict, memo);
                if (!subList.isEmpty()) {
                    for (String eachStr : subList) {
                        list.add(str + (eachStr.isEmpty() ? "" : " " + eachStr));
                    }
                }

            }
        }
        memo.put(s, list);
        return list;
    }
}